doc/dev/cephfs-mirroring.rst

```diff
@@ -0,0 +1,78 @@
+involves (efficiently) walking two (snapshot) trees. Since one of the snapshots is
+already synchronized, scanning the two snapshots locally would suffice. For the most
+part, identifying changed files between two snapshot does not involve querying the
+remote filesystem. However, there are cases where it is essential to query the remote
+filesystem:
+
+Failed Synchronization
+**********************
+
+Failure to transfer snapshot data (which could be due to variety of reasons) leaves
+the remote filesystem with incomplete data. Although retrying synchronization most
+likely overcomes transient issues, non-recoverable errors would need to be handled.
+When a snapshot synchronization hits a non-recoverable error, the snapshot is noted
+for failed synchronization and gets skipped. This is done either manually by the
+operator or when the number of consecutive retries (post failure) hits an configured
+limit.
+
+Skipping a "failed" snapshot leaves incomplete data on the remote filesystem. When the
+next snapshot is chosen for synchronization, local scanning of (two) snapshots is not
+feasible, since certain files (which were part of the "failed" snapshot) may have been
+transferred to the remote filesystem but do not exist in the "next" chosen snapshot.
+In such as case, scanning the two snapshots "locally" for changes would ignore the
+incomplete data on the remote filesystem left behind. In such as case identifying
+changes would need to compare files between the snapshot and the remote filesystem.
+Alternatively, the remote filesystem data could be purged followed by bulk transfer of
+data from the next chosen snapshot. Subsequent snapshots can then rely on scanning local
+snaphots for identifying changes to be transferred.
+
+Deleted Snapshots
+*****************
+
+Snapshots on the primary filesystem can be deleted even during synchronization.
+This needs to be handled by the mirror daemon and poses the same issues as mention
+above. Moreover, mirroring can be disabled in midst of snapshots being synchronized
+and enabled after deleting snapshots (including the one under synchronization) and
+poses the same set of issues.
+
+Snapshot Synchronization Order
+##############################
+
+Although the order in which snapshots get chosen for synhronization does not matter,
+it may be beneficial to pick snapshots based on creation order (using snap-id).
+
+Snapshot Incarnation
+####################
+
+A snapshot may be deleted and recreated (with the same name) with different contents.
+An "old" snapshot could have been syncrhonized (earlier) and the recreation of the
+snapshot could have been done when mirroring was disabled. Using snapshot names to
+infer the point-of-continuation would result in the "new" snapshot (incarnation)
+never getting picked up for synchronization.
+
+This can be mitigated by keeping the snap-id same for a snapshot in local and remote
```

Comment:
An alternate approach would be to maintain some form of global snap-id (perhaps a uuid) persisted in SnapInfo. This would require an interface to create a snapshot with a "known" uuid (when snapshotting remote filesystem subtree).

@ukernel thoughts?

Reply:
@vshankar 
Apart from establishing the uniqueness of a snap, how could the snap-id (uuid) be used to infer the chronological ordering of snap creation times ?

For the snap-replicator, the problem is to identify two consecutive snaps with a chronological ordering.
I know that a monotonically increasing integer could be of help.

Reply:
snap-ids are monotonically increasing integer values. The UUID proposal might not be space-efficient btw.

As far as establishing the point-of-continuation -- the global snap-id and (remote) snap-id would be used to identify the last synced snapshot.

Reply:
> An alternate approach would be to maintain some form of global snap-id (perhaps a uuid) persisted in SnapInfo. This would require an interface to create a snapshot with a "known" uuid (when snapshotting remote filesystem subtree).
> 
> @ukernel thoughts?

@ukernel ping?

Reply:
Including custom metadata (key-value pair in SnapInfo) could be of value and just might have more users. IIRC, mgr/volumes had the need to persist some timestamp when creating a snapshot. Such free-flowing metadata could be persisted in SnapInfo when creating a snapshot. However, this needs to be a separate interface for those who need to create snapshots and persist custom metadata.